src/fsevents/fsevents_stubs.c

```diff
@@ -137,11 +157,14 @@ static void dune_fsevents_callback(const FSEventStreamRef streamRef,
   }
   v_res = caml_callback_exn(t->v_callback, v_events_xs);
   if (Is_exception_result(v_res)) {
-    t->runloop->v_exn = Extract_exception(v_res);
-    CFRunLoopStop(t->runloop->runloop);
+    pthread_mutex_lock(&t->dq->dq_lock);
+    t->dq->v_exn = Extract_exception(v_res);
+    pthread_cond_broadcast(&t->dq->dq_finished);
+    pthread_mutex_unlock(&t->dq->dq_lock);
   }
   CAMLdrop;
   caml_release_runtime_system();
+  caml_c_thread_unregister();
```

Comment:
We are calling unregister in every single callback invocation. Are you sure this call is cheap enough for that?

Reply:
Is there a good metric for what you mean here? From what I can tell `caml_c_thread_register` creates a thread info block and then attaches it to an existing linked list of thread info blocks, and `caml_c_thread_unregister` reverses those changes. However, I'm not sure how expensive that is.

Reply:
I'm not so concerned about the linked list nor the block, but rather the lock that has to be acquired to do this. Given that this callback in our case isn't doing much work (just notifying which memoization nodes are out of date and hence invalidating the build can easily add up) acquiring this lock can easily affect the latency in watch mode.

I'd be more eager if this PR added some feature in exchange for the worse performance. Are there are any concrete benefits to using dispatch queues that I'm unaware of perhaps?

Reply:
The original reason is that [`FSEventStreamScheduleWithRunLoop` is deprecated](https://developer.apple.com/documentation/coreservices/1447824-fseventstreamschedulewithrunloop) with a suggestion to use [`FSEventStreamSetDispatchQueue`](https://developer.apple.com/documentation/coreservices/1444164-fseventstreamsetdispatchqueue) instead. I think in practice, since we're using a serial queue, the background thread running the callback should remain the same, but I couldn't find an easy way to manually manage the threads used for the queue, which could allow us to only run `caml_c_thread_register` as needed.

Alternatively, we could use [`dispatch_get_main_queue`](https://developer.apple.com/documentation/dispatch/1452921-dispatch_get_main_queue), which uses a serial dispatch queue in the main thread. This would have roughly the same behavior as the original code that use [`CFRunLoopGetCurrent`](https://developer.apple.com/documentation/corefoundation/1542428-cfrunloopgetcurrent) assuming there was originally only one thread, but I think we would probably need to run the main dispatch queue somehow with either [`dispatch_main`](https://developer.apple.com/documentation/dispatch/1452860-dispatch_main) or [`CFRunLoopRun`](https://developer.apple.com/documentation/corefoundation/1542011-cfrunlooprun).

Reply:
As an alternative, I've introduced some thread-local storage that keeps track of whether `caml_c_thread_register` has been called already by the same thread, and only unregisters once the thread exits. Since a serial dispatch queue often uses the same thread, this should in principle reduce the number of register/unregister calls needed, but additional testing is probably needed.

Reply:
Yeah, that's a bit better. If you could do some testing measuring that we aren't spamming caml_c_thread_register/unregister then this should be good to go.

Another alternative that's a little more bulletproof but requires a bit more code is to have the code in the dispatch queue populate some sort of queue with events, and require the OCaml side to poll this queue for events in a loop. This will require some synchronization and the use of CV's to avoid busy polling, but it's probably the best we can do here.

cc @patricoferris who discussed this issue with me.

Reply:
@rgrinberg Sorry just got some time to come back to look at this. I can confirm that running the `test/blackbox-tests/test-cases/watching` test that the same background thread is being reused and `caml_c_thread_register` is only being run once. Are there some longer workflows that I can test to make sure the behavior is also as expected?

Reply:
You could try running longer builds on more serious projects. But your test seems enough to me.

Reply:
If you have some spare cycles, I would suggest looking at the queue based workaround I suggest previous comments. I think that should guarantee this is well behaved.